File: BT_Right_side.py
# Time Complexity:O(n) (where n = number of nodes)
# Space Complexity:O(h) (where h = height of the tree)
# Did this code successfully run on LeetCode?:Yes
# Problems faced while coding this:None
# Approach: DFS approach
# here the right side view is acheived by recursing on left side.

class Solution:
    
    def rightSideView(self, root: TreeNode) -> List[int]:
        self.result = []
        if root == None:
            return self.result
        self.dfs(root, 0)
        return self.result

    def dfs(self, root, lvl):
        if root == None:
            return 
        if lvl == len(self.result):
            self.result.append(root.val)
        else:
            self.result[lvl] = root.val

        self.dfs(root.left, lvl + 1)
        self.dfs(root.right, lvl + 1)

File: BTRightSide.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 
 TC = O(N) where N is the number of nodes in the binary tree
 SC = O(H) where H is the height of the binary tree.

 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        
        List<Integer> result = new ArrayList<>();
        
        if (root == null)
        {
            return result;
        }
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        
        while(!queue.isEmpty())
        {
            int qSize = queue.size();
            
            for (int i=0; i<qSize; i++)
            {
                TreeNode remove = queue.remove();
                
                if (remove.left != null)
                {
                    queue.add(remove.left);
                }
                
                if (remove.right != null)
                {
                    queue.add(remove.right);
                }
                
                if (i == qSize - 1)
                {
                    result.add(remove.val);
                }
            }
        }
        
        return result;
    }
}
